upcoming: [DI-27318] - Dimension Filter support for Object Storage Alerts

santoshp210-akamai opened the pull request:
## Description 📝
 - Support for custom fetching for the Endpoint Dimension Filter in Object Storage Alerts
 - Minor changes and separation of logic in ValueFieldRenderer component

## Changes  🔄

List any change(s) relevant to the reviewer.

- Standalone DimensionFilterValue components (`DimensionFilterAutocomplete`, `FirewallDimensionFilterAutocomplete`, `ObjectStorageDimensionFilterAutocomplete`) depending on the use-case ( Avoids fetching irrelevant hooks for some serviceTypes)
- New `useObjectStorageFetchOptions `hook for endpoints dimension filter
- Utils for scope related filtering for object storage endpoints dimension filter
- Unit Tests for the components and utils
- Unified Interfaces for DimensionFilterAutocompleteProps and FetchOptionsProps

### Scope 🚢

 Upon production release, changes in this PR will be visible to:

- [ ] All customers
- [x] Some customers (e.g. in Beta or Limited Availability)
- [ ] No customers / Not applicable

## Target release date 🗓️

1st October

## Preview 📷

**Include a screenshot `<img src="" />` or video `<video src="" />` of the change.**

:lock: Use the [Mask Sensitive Data](https://cloud.linode.com/profile/settings) setting for security.

:bulb: For changes requiring multiple steps to validate, prefer a video for clarity.

| Before  | After   |
| ------- | ------- |
| 📷 | 📷 |

## How to test 🧪

### Prerequisites

(How to setup test environment)

- ...
- ...

### Reproduction steps

(How to reproduce the issue, if applicable)

- [ ] ...
- [ ] ...

### Verification steps

(How to verify changes)

- [ ] ...
- [ ] ...

<details>
<summary> Author Checklists </summary>

## As an Author, to speed up the review process, I considered 🤔

*Check all that apply*

- [ ] Use React components instead of HTML Tags
- [ ] Proper naming conventions like cameCase for variables & Function & snake_case for constants
- [ ] Use appropriate types & avoid using "any"
- [ ] No type casting & non-null assertions
- [ ] Adding a [changeset](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md#writing-a-changeset)
- [ ] Providing/Improving test coverage
- [ ] Use sx props to pass styles instead of style prop
- [ ] Add JSDoc comments for interface properties & functions
- [ ] Use strict equality (===) instead of double equal (==)
- [ ] Use of named arguments (interfaces) if function argument list exceeds size 2
- [ ] Destructure the props
- [ ] Keep component size small & move big computing functions to separate utility
- [ ] 📱 Providing mobile support

<br/>

- [x] I have read and considered all applicable items listed above.

## As an Author, before moving this PR from Draft to Open, I confirmed ✅

- [x] All tests and CI checks are passing
- [x] TypeScript compilation succeeded without errors
- [x] Code passes all linting rules

</details>

venkymano-akamai commented on packages/manager/src/features/CloudPulse/Alerts/CreateAlert/Criteria/DimensionFilterValue/ValueFieldRenderer.tsx:
```diff
@@ -156,24 +139,56 @@ export const ValueFieldRenderer = (props: ValueFieldRendererProps) => {
-      />
-    );
+
+    if (config.useCustomFetch === 'firewall')
```
@santoshp210-akamai , should we go for a map based approach? instead of this if else since this will grow in size?, lets check that way as well

santoshp210-akamai replied:
@venkymano-akamai , Currently there are only two special cases and even if we use map based approach it will grow because of props. I think having a prop or component builder for future might solve the growing issue rather than map based approach to render components.

nikhagra-akamai commented on packages/manager/src/features/CloudPulse/Alerts/CreateAlert/Criteria/DimensionFilterValue/DimensionFilterAutocomplete.tsx:
```diff
@@ -64,28 +20,35 @@ export const DimensionFilterAutocomplete = (
+  const options = useMemo(
+    () =>
+      getStaticOptions(
+        serviceType ?? undefined,
```
```suggestion
        serviceType,
```

lets convert the serviceType parameter to null instead of undefined here. Same undefined to null conversion is happening so its better keep it null instead of undefined [utils.ts](https://github.com/ACLPManager/manager/blob/44ac5ec199296e2fd1c4c9abb0a953f4a3de639e/packages/manager/src/features/CloudPulse/Alerts/CreateAlert/Criteria/DimensionFilterValue/utils.ts#L81)

nikhagra-akamai commented on packages/manager/src/features/CloudPulse/Alerts/CreateAlert/Criteria/DimensionFilterValue/DimensionFilterAutocomplete.tsx:
```diff
@@ -64,28 +20,35 @@ export const DimensionFilterAutocomplete = (
+      getStaticOptions(
+        serviceType ?? undefined,
+        dimensionLabel ?? '',
+        values ?? []
```
```suggestion
        values ?? []
```
since you are sending as empty fallback value so [this](https://github.com/ACLPManager/manager/blob/44ac5ec199296e2fd1c4c9abb0a953f4a3de639e/packages/manager/src/features/CloudPulse/Alerts/CreateAlert/Criteria/DimensionFilterValue/utils.ts#L83) will never be null, so its better we update it to avoid returning fallbacks

nikhagra-akamai commented on packages/manager/src/features/CloudPulse/Alerts/CreateAlert/Criteria/DimensionFilterValue/DimensionFilterAutocomplete.tsx:
```diff
@@ -64,28 +20,35 @@ export const DimensionFilterAutocomplete = (
       label="Value"
       limitTags={1}
-      loading={!disabled && isLoading && !isError}
+      loading={!disabled}
```
why just !disabled? that means every time this component will be in enabled state always have loading. Control loading state is with proper loading variable

santoshp210-akamai replied:
@nikhagra-akamai , Yeah this is a miss on my end. The loading state is technically never possible since these options are from the metric-definitions call. Removing the loading state.

nikhagra-akamai commented on packages/manager/src/features/CloudPulse/Alerts/CreateAlert/Criteria/DimensionFilterValue/FirewallDimensionFilterAutocomplete.tsx:
```diff
@@ -0,0 +1,69 @@
+    scope,
+  });
+  return (
+    <Autocomplete
+      data-qa-dimension-filter={`${name}-value`}
+      data-testid="value"
+      disabled={disabled}
+      errorText={
+        errorText ?? (isError ? 'Failed to fetch the values.' : undefined)
+      }
+      isOptionEqualToValue={(option, value) => value.value === option.value}
+      label="Value"
+      limitTags={1}
+      loading={!disabled && isLoading && !isError}
+      multiple={multiple}
+      onBlur={fieldOnBlur}
+      onChange={(_, selected, operation) => {
+        const newValue = handleValueChange(
+          selected,
+          operation,
+          multiple ?? false
+        );
+        fieldOnChange(newValue);
+      }}
+      options={values}
+      placeholder={placeholderText}
+      sx={{ flex: 1 }}
+      value={resolveSelectedValues(values, fieldValue, multiple ?? false)}
+    />
+  );
```
can't we use DimensionFilterAutoComplete instead of defining new autocomplete, that is kinda similar?

venkymano-akamai replied:
@santoshp210-akamai , thinking of similar approach, lets have different components for object storage and firewall, but reuse the autocomplete by passing props from the components of object storage and firewall?

santoshp210-akamai replied:
@nikhagra-akamai , The idea was that  DimensionFilterAutoComplete will be used where we depend on the metric-definitions API to have the options. 

If we need to use DimensionFilterAutoComplete. We need to have some logic of conditionally calling the getStaticOptions util .This makes the component coupled. The main idea was to isolate the components with their logic to themselves.

nikhagra-akamai commented on packages/manager/src/features/CloudPulse/Alerts/CreateAlert/Criteria/DimensionFilterValue/ObjectStorageDimensionFilterAutocomplete.tsx:
```diff
@@ -0,0 +1,72 @@
+    serviceType,
+  });
+
+  return (
+    <Autocomplete
+      data-qa-dimension-filter={`${name}-value`}
+      data-testid="value"
+      disabled={disabled}
+      errorText={
+        errorText ??
+        (isError ? 'Failed to fetch Object Storage endpoints.' : undefined)
+      }
+      isOptionEqualToValue={(option, value) => value.value === option.value}
+      label="Value"
+      limitTags={1}
+      loading={!disabled && isLoading && !isError}
+      multiple={multiple}
+      onBlur={fieldOnBlur}
+      onChange={(_, selected, operation) => {
+        const newValue = handleValueChange(
+          selected,
+          operation,
+          multiple ?? false
+        );
+        fieldOnChange(newValue);
+      }}
+      options={values}
+      placeholder={placeholderText}
+      sx={{ flex: 1 }}
+      value={resolveSelectedValues(values, fieldValue, multiple ?? false)}
+    />
```
same for this autocomplete

nikhagra-akamai commented on packages/manager/src/features/CloudPulse/Alerts/CreateAlert/Criteria/DimensionFilterValue/ValueFieldRenderer.tsx:
```diff
@@ -156,24 +139,56 @@ export const ValueFieldRenderer = (props: ValueFieldRendererProps) => {
-      />
-    );
+
+    if (config.useCustomFetch === 'firewall')
```
just use switch instead of if-else ladder, that will be more clean and easy to add new filter

santoshp210-akamai commented:
@nikhagra-akamai , @venkymano-akamai . Addressed the comments. With respect to this specific comment https://github.com/ACLPManager/manager/pull/481#discussion_r2387737411. I have discussed with @venkymano-akamai and there is no useful impact from using DimensionFilterAutocomplete instead of Autocomplete component. We still use the same props and we also have benefit of isolating logic to each of the customization.
